Add image comparison features to details dialog

glubsy opened the pull request:
* Add splitter in order to hide the details table.
* Add a toolbar to the Details Dialog window to allow for better image
comparisons: zoom in/out, swap pixmaps in place, best-fit-to-viewport.
Scrollbars and viewports are synchronized.

Notes about the patch:
* We only use the ScrollAreaImageViewer class, and the ScrollAreaController here, the other image viewer classes (QWidgetImageViewer and GraphicsViewImageViewer) and their corresponding controllers (QWidgetController and GraphicsViewController) have been left in in case we need them later and for study purposes.
* The QWidget-based image viewer is the simplest implementation of all, but lacks scrollbars and also has some issues when scaling up and panning around. Probably something fixable but I have not pursued it further.
* The GraphicsView-based image viewer is more unnecessarily full-fledged at this point and I decided not to used it. The ScrollArea version is more than enough and works slightly better too (there is some minor margin problem with the GraphicsView version which I have not worked on fixing) 
* The QScrollArea-based image viewer is the most balanced of the three and works just as well while not carrying all the "bloat" of the QGraphicsView widget-based implementation.

[Screencast video demonstration of the implemented features. ](https://open.lbry.com/@Fuan_Karuru:e/demo_detailsdialog:8?r=7EFySAboE3WV7NfoVfWbJfqdYamt5hMe)

glubsy committed:
add zoom and swap buttons to details dialog

glubsy committed:
attempt at double click on Qlabel

glubsy committed:
add normal size button

glubsy committed:
drag mouse with ImageViewer class

glubsy committed:
working zoom functions, mouse wheel event

glubsy committed:
working synchronized panning

glubsy committed:
only update delta when mouse is being dragged to reduce paint events

glubsy committed:
Working synchronized views.

glubsy committed:
Controller class to decouple from the dialog class

The controller singleton acts as a proxy to relay
signals from each widget to the other
It should help encapsulating things better if we need to
use a different class for image viewers in the future.

glubsy committed:
Disable unused methods from controller

* setPixmap() now disables the QWidget automatically if the pixmap passed is null.
* the controller relays repaint events to the other widget

glubsy committed:
Use native QPixmap swap() method instead of manual setPixmap()

When swapping images, use getters to hopefully get a reference to
each pixmap and swap them within a single slot.

glubsy committed:
Implement mostly working ScrollArea viewer

Using a QWidget inside the QScrollArea mostly works
but we only move around the pixmap inside the QWidget,
not the QWidget itself, which doesn't update scrollbars.
Need a better implementation.

glubsy committed:
Add mostly working ScrollArea imge viewers

* Work around flickering of scrollbars due to
GridLayout resizing on odd pixels by disabling
the scrollbars while BestFit is active
* Also setting minimum column width to work around
the issue above.
* Avoid updating scrollbar values twice by using a
simple boolean lock

glubsy committed:
Work around resizing down offset by 1 pixel.

glubsy committed:
Add QSplitter to hide TableView in DetailsDialog

glubsy committed:
Keep scale accross files of the same dupe group.

* Also fix scaled down pixmap when updating pixmap in the same group
* Fix ignoring mouse wheel event when max scale has been reached
* Fix toggle scrollbars when asking for normal size

glubsy committed:
Fix view resetting when selecting reference only.

* Needed to ignore the scrollbar changes in the disabled
panel, sine a null pixmap would reset the bars to 0 and affect
the selected viewer.
* Keep view as same scale accross entries from the same group.

glubsy committed:
Add working zoom functions to GraphicsView viewers.

glubsy committed:
Move buttons into the toolbar class.

* Moved the QToolbar into the image viewer's  translation unit.
* QAction are still attached to the dialog window for shortcuts to work

glubsy committed:
Cleanup of details table

glubsy committed:
Add image comparison features to details dialog

* Add splitter in order to hide the details table.
* Add a toolbar to the Details Dialog window to allow for better image
comparisons: zoom in/out, swap pixmaps in place, best-fit-to-viewport.
Scrollbars and viewports are synchronized.

glubsy committed:
Fix flake8 minor issues

glubsy committed:
Merge commit 'b0a256f0'

glubsy committed:
Make details dialog dockable

glubsy committed:
Fix ME/SE details dialogs, add preferences

* Fix ME and SE versions of details dialog not displaying their content properly after change to QDockWidget
* Add option to toggle titlebar and orientation of titlebar in preferences dialog
* Fix setting layout on PE details dialog window while layout already set, by removing the self (parent) reference in constructing the QSplitter

glubsy committed:
Fix scrollbar displayed while splitter maxed out

* For some reason the table's height is a few pixel longer on Windows so we work around the issue by adding a small offset to the maximum height hint.
* No idea about MacOS yet but this might need the same treatment.

glubsy committed:
Add custom icons

* Use custom icons on platforms which do not provide theme
* Old zoom icons credits to "schollidesign" from icon pack Office and Entertainment (GPL licence).
* Exchange icon credit to Jason Cho (Unknown license).
* Use hack to resize viewers on first show() as well

glubsy committed:
Add credit for icons used, upscale exchange icon

* Jason Cho gave his express permission to use the icon (it was made 10 years ago and he doesn't have the source files anymore)
* Used waifu2x to upscale the icon
* Used GIMP to draw dark outline around the icon
* Source files are included

glubsy committed:
Merge branch 'dockable_windows' into details_dialog_improvements_dev

glubsy committed:
Fix computing and setting offset to 0 for tableview

glubsy committed:
Prevent Windows from floating if no decoration

* Windows users cannot move a window which has no native decorations. Toggling a dock widget's titlebar off also removes native decorations on a floating window. Until we implement a replacement titlebar by overriding paintEvents, simply force the floating window to go back to docked state after we toggled the titlebar off.

glubsy committed:
Update preferences on show(), not in constructor

* If the dialog window shouldn't have a titlebar during construction, update accordingly only when showing to fix Windows displaying a window without titlebar on first show
* Only save geometry if the window is floating. Otherwise geometry while docked is saved whih gives weird results on subsequent starts, since it may be floating by default anyway (at least on Linux where titlebar being disabled is allowed while floating)
* Vertical title bar doesn't seem to work on Windows, add note in preferences dialog

glubsy commented:
Here is a [quick screencast demonstrating the dockable widget feature](https://open.lbry.com/@Fuan_Karuru:e/2020-07-15_23-11-28_edited:8?r=7EFySAboE3WV7NfoVfWbJfqdYamt5hMe). 
Should work on Windows as demonstrated [here](https://open.lbry.com/@Fuan_Karuru:e/2020-07-16_00-02-11:6?r=7EFySAboE3WV7NfoVfWbJfqdYamt5hMe). Haven't tested on MacOS (such a PAIN to setup a virtual machine that works with that malware).
I am not going to develop this feature in a separate branch anymore as it is getting messy keeping track of all changes.

glubsy committed:
Prevent zoom for images of differing dimensions

* If images are not the same size, prevent zooming features from being used by disabling the normal size button, only enable swap

glubsy committed:
Fix resize of top frame not updating scaled pixmap

* Also limit viewing features such as zoom levels when files have different dimensions
* GraphicsViewImageViewer is still a bit buggy: the scrollbars are toggled on when the pixmap is null in the reference viewer (we do not use that class right anyway)

glubsy committed:
Squashed commit of the following:

commit ac941037ff51158b64daa65c244df26346af10cf
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Thu Jul 16 22:21:24 2020 +0200

    Fix resize of top frame not updating scaled pixmap

    * Also limit viewing features such as zoom levels when files have different dimensions
    * GraphicsViewImageViewer is still a bit buggy: the scrollbars are toggled on when the pixmap is null in the reference viewer (we do not use that class right anyway)

commit 733b3b0ed4fbd6de908c968402af03879df3336f
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Thu Jul 16 01:31:24 2020 +0200

    Prevent zoom for images of differing dimensions

    * If images are not the same size, prevent zooming features from being used by disabling the normal size button, only enable swap

commit 9168d72f38faaf0a12230cd544f14190cd29fca4
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Wed Jul 15 22:47:32 2020 +0200

    Update preferences on show(), not in constructor

    * If the dialog window shouldn't have a titlebar during construction, update accordingly only when showing to fix Windows displaying a window without titlebar on first show
    * Only save geometry if the window is floating. Otherwise geometry while docked is saved whih gives weird results on subsequent starts, since it may be floating by default anyway (at least on Linux where titlebar being disabled is allowed while floating)
    * Vertical title bar doesn't seem to work on Windows, add note in preferences dialog

commit 75621cc816120597f493e0debc6d88e2e0bbd30a
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Wed Jul 15 22:04:19 2020 +0200

    Prevent Windows from floating if no decoration

    * Windows users cannot move a window which has no native decorations. Toggling a dock widget's titlebar off also removes native decorations on a floating window. Until we implement a replacement titlebar by overriding paintEvents, simply force the floating window to go back to docked state after we toggled the titlebar off.

commit 3c816b2f11ddc66a78cdc6327ee102df46d1a552
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Wed Jul 15 21:43:01 2020 +0200

    Fix computing and setting offset to 0 for tableview

commit 85d6e05cd406b999e8f6ae421a9746a0c9f146bb
Merge: 66127d02 3eddeb6a
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Wed Jul 15 21:25:44 2020 +0200

    Merge branch 'dockable_windows' into details_dialog_improvements_dev

commit 66127d025e9a497ee13126f955166946acdb35a8
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Wed Jul 15 20:22:13 2020 +0200

    Add credit for icons used, upscale exchange icon

    * Jason Cho gave his express permission to use the icon (it was made 10 years ago and he doesn't have the source files anymore)
    * Used waifu2x to upscale the icon
    * Used GIMP to draw dark outline around the icon
    * Source files are included

commit 58c675d1fa90a7247233d9887a460cf5a8e4cbf5
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Wed Jul 15 05:25:47 2020 +0200

    Add custom icons

    * Use custom icons on platforms which do not provide theme
    * Old zoom icons credits to "schollidesign" from icon pack Office and Entertainment (GPL licence).
    * Exchange icon credit to Jason Cho (Unknown license).
    * Use hack to resize viewers on first show() as well

commit 95b8406c7b97aab170d127b466ff506b724def3c
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Wed Jul 15 04:14:24 2020 +0200

    Fix scrollbar displayed while splitter maxed out

    * For some reason the table's height is a few pixel longer on Windows so we work around the issue by adding a small offset to the maximum height hint.
    * No idea about MacOS yet but this might need the same treatment.

commit 3eddeb6aebc99126e62eb05af60333ba3bd22e82
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Tue Jul 14 17:37:48 2020 +0200

    Fix ME/SE details dialogs, add preferences

    * Fix ME and SE versions of details dialog not displaying their content properly after change to QDockWidget
    * Add option to toggle titlebar and orientation of titlebar in preferences dialog
    * Fix setting layout on PE details dialog window while layout already set, by removing the self (parent) reference in constructing the QSplitter

commit 56912a71084415eac2f447650279d833d9857686
Author: glubsy <glubsy@users.noreply.github.com>
Date:   Mon Jul 13 05:06:04 2020 +0200

    Make details dialog dockable

glubsy committed:
Add tabs to preference dialog.

glubsy committed:
Fix Restore Default Preferences button

* When clicking the "Restore Default" in the preferences dialog, only affect the preferences displayed in the current tab. The hidden tab should not be affected by this button.

glubsy committed:
Add option to use internal icons in details dialog

* On Windows and MacOS, no idea how themes work so only allow Linux to use their theme icons
* Internal icons are used by default on non-Linux platforms

glubsy committed:
Add color picker buttons to preferences dialog

* Buttons display the color currently in use
* Result table uses selected colors accordingly
* Keep items aligned with GridLayouts in preference dialog
* Reordering of items in a more logical manner*

glubsy commented:
Added some options to the preference dialog all the while improving it a bit:
* Separated general options from display options
* Added color pickers to override the default colors for reference and delta highlighting colors (which were hardcoded)
* The "Restore defaults" button restores the default values only for the currently displayed tab
* Will add option to select color for diff in the details dialog after that other branch has been merged
![2020-07-22_22-15-13_](https://user-images.githubusercontent.com/12895548/88226294-20614000-cc6c-11ea-94a1-f3727d4ba76d.gif)

glubsy committed:
Merge branch 'preferences_tabs' into details_dialog_improvements

glubsy commented:
I am done working on these branches for now, I consider them stable. Feel free to review.

arsenetar commented:
Well, this looks pretty good, I think the other merge is conflicting with this.  I need to play around with this a bit more, it seemed like the docking did not work until I toggled the option on/off on the settings, then it seemed to work.  Also toggling the details view via the menu or shortcut key appears changes it from the non-dockable version and back.  I expected that to actually hide/show it.  I need to play around with the rest of it a bit more.

glubsy commented:
Right, those minor conflicts were expected and I'll fix them. 
I will have a closer look at the other problem too.

glubsy committed:
Merge branch 'master' into details_dialog_improvements

glubsy committed:
Fix swapping table view columns

We now have only two columns to swap, not 3.

glubsy committed:
Fix toggling details view via menu or shortcut

* Using Ctrl+I would toggle the title bar on/off

glubsy commented:
I could make the Ctrl+I and "Details" button shortcuts simply close and reopen the panel if you prefer. 

Normally the window would be floating by default and should have a title bar which allows dragging and docking.
I suspect the initial dock not working problem might be from qt/details_dialog.py:44 in update_options(). 
Edit: there is indeed a small bug when toggling the titlebar while the window had already been created, will look into it.

glubsy commented:
Latest commit should take care of this bug.

glubsy committed:
Fix title bar toggling on/off when dialog

glubsy committed:
Toggle visibility of details dialog

* When using the Ctrl+I shortcut or the "Details" button in the Results window, toggle the details dialog on/off.
* This works also while it is docked.

glubsy committed:
Add color preference for delta in details table

glubsy committed:
Add preference to turn off scrollbars in viewers

Refactor preference Display page to only include PE specific preferences in the PE mode.

glubsy committed:
Fix crash quitting while details dialog active

* While the details dialog is opened, if quit is triggered, the error message "'DetailsPanel' object has no attribute '_table'" is reported
* A workaround is to cleanly close the dialog before tear down

glubsy committed:
Stretch last header section in Result window

glubsy committed:
Avoid crash when quitting

* If details dialog failed to be created for some reason, avoid crashing by dereferencing a null pointer

glubsy committed:
Remove unused properties in details table headers

glubsy committed:
Save dock panel position on quit

* Restore the details dialog dock position if it was previously docked (i.e. not floating).
* Since the details_dialog instance was not deleted after closing by default, the previous instances were still saving their own geometry. We now delete them explicitely if we have to recreate a new instance to avoid the signal triggering the callback to save the geometry.
* Since restoreGeometry() and saveGeometry() are only called in our QDockWidget, it should be safe to modify the methods for the Preferences class (in qtlib).

glubsy committed:
Fix layout error

* Avoid attempting to add a QLayout to DetailsDialog which already has a layout by removing superfluous layout setup.

glubsy commented:
Seems to work fine in MacOS Catalina:
![simplescreenrecorder-2020-08-01_01 54 00](https://user-images.githubusercontent.com/12895548/89088789-61f39880-d39a-11ea-829f-9a7c49933a43.gif)

glubsy committed:
Fix 1 pixel sized color in color picker buttons

* On Linux, even with 1 pixel size, the button is filled entirely with the color selected
* On MacOS, the color pixmap stays at 1 pixel so we hard code the size to 16x16

arsenetar commented:
@glubsy, the tab PR is merged so you can work out the conflicts here when you get a chance.

glubsy committed:
Tweak preference layout for better readability

* We use GroupBoxes to group items together and surround them in a frame
* Remove separator lines to avoid cluttering
* Adjust columns and their stretch factors for better alignment of buttons

glubsy committed:
Use FormLayout instead of GridLayout

QFormLayout should adhere to each platform's style better. It also simplifies the code a bit since we don't have to setup the labels, etc.

glubsy commented:
The display tab in the preference dialog should look slightly better now:
![2020-08-01_18-54-33](https://user-images.githubusercontent.com/12895548/89106166-dcb6c500-d427-11ea-91a6-2e3f53b839e9.png)
![2020-08-01_18-55-07](https://user-images.githubusercontent.com/12895548/89106171-eccea480-d427-11ea-957f-2201e3f66d1f.png)
Edit: updated versions

glubsy committed:
Merge branch 'master' into details_dialog_improvements

glubsy committed:
Revert stretching last section in Result window

* It seems that stretching the last section automatically is a bit inconvenient on MacOS as it will grow beyond the window border.
* Keep it as it was before for now until a better solution is devised.

glubsy committed:
Group general interface options together

* Use QGroupBox to keep items together on the display tab in the preference dialog just like for the other options.
* It is probably not be necessary to keep these as class members

glubsy commented:
I think we're good here, not much to add anymore for now. We might need to start considering releasing 4.0.5 soon. ;)

glubsy committed:
Add stretch to bottom of General pref tab

glubsy committed:
Tweak visuals for cache selection item

glubsy committed:
Fix alignment for combo box's label

glubsy committed:
Prevent Directories tab from closing on MacOS

* The close button on custom tabs cannot be hidden on MacOS for some reason.
* Prevent the directories tab from closing if the close button was clicked by mistake

glubsy committed:
Fix adding new Result tab if already existed

* Whenever the Result Window already existed and its tab was in second position, and if the ignore list tab was in 3rd position, asking to show the Result window through the View menu would add a new tab and push the Result tab to the third position (ignore list tab would then become 2nd position).
* Fix view menu Directories entry not switching to index "0" in custom tab bar.

glubsy committed:
Fix typo in class member reference

arsenetar commented on qtlib/preferences.py:
```diff
@@ -123,16 +124,22 @@ def saveGeometry(self, name, widget):
         # We save geometry under a 5-sized int array: first item is a flag for whether the widget
```
Probably should update this comment to reflect the new values that have been added.

glubsy replied:
Right forgot about that change, thanks.

arsenetar commented on qt/se/details_dialog.py:
```diff
@@ -19,11 +19,14 @@ def _setupUi(self):
-        self.verticalLayout.setContentsMargins(0, 0, 0, 0)
+        # self.verticalLayout = QVBoxLayout()
+        # self.verticalLayout.setSpacing(0)
+        # self.verticalLayout.setContentsMargins(0, 0, 0, 0)
```
Probably just remove this commented lines, always have git history.

arsenetar commented on qt/se/details_dialog.py:
```diff
@@ -19,11 +19,14 @@ def _setupUi(self):
-        self.verticalLayout.addWidget(self.tableView)
+        # self.verticalLayout.addWidget(self.tableView)
+        # self.centralWidget = QWidget()
+        # self.centralWidget.setLayout(self.verticalLayout)
```
Can remove these as well.

arsenetar commented on qt/me/details_dialog.py:
```diff
@@ -19,11 +19,14 @@ def _setupUi(self):
-        self.verticalLayout.setContentsMargins(0, 0, 0, 0)
+        # self.verticalLayout = QVBoxLayout(self)
+        # self.verticalLayout.setSpacing(0)
+        # self.verticalLayout.setContentsMargins(0, 0, 0, 0)
```
Can remove these commented lines, git history will be there.

arsenetar commented on qt/me/details_dialog.py:
```diff
@@ -19,11 +19,14 @@ def _setupUi(self):
-        self.verticalLayout.addWidget(self.tableView)
+        # self.verticalLayout.addWidget(self.tableView)
+        # self.centralWidget = QWidget()
+        # self.centralWidget.setLayout(self.verticalLayout)
```
Also remove these.

glubsy committed:
Don't show parent window's context menu on viewers

* When right clicking on image viewers while they are docked, the context menu of the Results window showed up.
* This also enables capture of right click and middle click buttons to drag around images, which solves a conflict with some theme engines that enable left mouse button click to drag a window's position regardless of where the event happens, hence blocking the panning.
* Probably unnecessary to check which button is released.

arsenetar commented on qt/dg.qrc:
```diff
@@ -5,5 +5,10 @@
+    <file alias="zoom_out">../images/old_zoom_out.png</file>
+    <file alias="zoom_original">../images/old_zoom_original.png</file>
+    <file alias="zoom_best_fit">../images/old_zoom_best_fit.png</file>
 </qresource>
```
Commenting here, but what is the reason for all the other unused images added here?  These 5 seem to be the only ones used here.

glubsy replied:
Indeed those are the original files from which I derive the final icon files. These should not be included in the release package, but should probably still be included in the source tar balls. Since the source files were lost by the author (Jason Cho), I upscaled them (with waifu2x) but left the original files in case we need them later. Turns out I'll modify that icon right now.

arsenetar replied:
Ok

arsenetar commented on qt/pe/details_dialog.py:
```diff
@@ -4,115 +4,147 @@
+
+        # This works when expanding but it's ugly:
+        if self.selectedImageViewer.size().width() > self.referenceImageViewer.size().width():
+            # print(f"""Before selected size: {self.selectedImageViewer.size()}\n""",
```
Clean up these print debugging statements.

arsenetar commented on qtlib/about_box.py:
```diff
@@ -69,6 +69,21 @@ def _setupUi(self):
+        self.label_4.setWordWrap(True)
+        self.label_4.setTextFormat(Qt.RichText)
+        self.label_4.setOpenExternalLinks(True)
+        self.label_4.setText(tr(
```
This is the first time something like this has been done in this project, most other non-code contributes have been in CREDITS.  Granted the GPL ones are a bit different than that intention.  Not 100% sure on doing it this way.

glubsy replied:
I had overlooked the CREDITS file and I was so invested in displaying them there. Will remove.

glubsy replied:
However, maybe it would be nice to display the whole credits in that dialog box later?

arsenetar replied:
Yeah I don't think I am necessarily opposed to that, although I think it needs to be thought of how to do it in a way that would be "fair".   Since those making code contributes are not in credits only the non-code items.  I think if we were going to do something like that it probably should include all the contributions.  I could see that being something to add post 4.0.5.

glubsy replied:
Right, I will have to gather all contributors to the git repo (not hat difficult). I could make a tab for code contributors and maintainers, another tab for translations, another tab for everything else. Will keep a note about this in my TODO list for later.

arsenetar reviewed (commented):
Reviewed these changes, had a few minor items.

glubsy committed:
Improve comment reflecting modification of function

glubsy committed:
Remove obsolete UI setup code

glubsy committed:
Remove print debug statements

glubsy committed:
Remove icon credits from about box

* Moved credits to CREDITS file
* Updated exchange icon with higher hue contrast for better visibility on dark backgrounds
